src/storage/v2/inmemory/edge_type_property_index.cpp

```diff
@@ -0,0 +1,333 @@
+    : self_(self),
+      index_iterator_(index_iterator),
+      current_edge_accessor_(EdgeRef{nullptr}, EdgeTypeId::FromInt(0), nullptr, nullptr, self_->storage_, nullptr),
+      current_edge_(nullptr) {
+  AdvanceUntilValid();
+}
+
+InMemoryEdgeTypePropertyIndex::Iterable::Iterator &InMemoryEdgeTypePropertyIndex::Iterable::Iterator::operator++() {
+  ++index_iterator_;
+  AdvanceUntilValid();
+  return *this;
+}
+
+void InMemoryEdgeTypePropertyIndex::Iterable::Iterator::AdvanceUntilValid() {
+  for (; index_iterator_ != self_->index_accessor_.end(); ++index_iterator_) {
+    auto *from_vertex = index_iterator_->from_vertex;
+    auto *to_vertex = index_iterator_->to_vertex;
+
+    if (!IsEdgeVisible(index_iterator_->edge, self_->transaction_, self_->view_) || from_vertex->deleted ||
+        to_vertex->deleted) {
+      continue;
+    }
+
+    const bool edge_was_deleted = index_iterator_->edge->deleted;
+    auto [edge_ref, edge_type, deleted_from_vertex, deleted_to_vertex] = GetEdgeInfo();
+    MG_ASSERT(edge_ref != EdgeRef(nullptr), "Invalid database state!");
+
+    if (edge_was_deleted) {
+      from_vertex = deleted_from_vertex;
+      to_vertex = deleted_to_vertex;
+    }
+
+    auto accessor = EdgeAccessor{edge_ref, edge_type, from_vertex, to_vertex, self_->storage_, self_->transaction_};
+    if (!accessor.IsVisible(self_->view_)) {
+      continue;
+    }
+
+    current_edge_accessor_ = accessor;
+    current_edge_ = edge_ref;
+    break;
+  }
+}
+
+std::tuple<EdgeRef, EdgeTypeId, Vertex *, Vertex *> InMemoryEdgeTypePropertyIndex::Iterable::Iterator::GetEdgeInfo() {
+  auto *from_vertex = index_iterator_->from_vertex;
+  auto *to_vertex = index_iterator_->to_vertex;
+
+  if (index_iterator_->edge->deleted) {
+    const auto missing_in_edge =
+        VertexDeletedConnectedEdges(from_vertex, index_iterator_->edge, self_->transaction_, self_->view_);
+    const auto missing_out_edge =
+        VertexDeletedConnectedEdges(to_vertex, index_iterator_->edge, self_->transaction_, self_->view_);
+    if (missing_in_edge && missing_out_edge &&
+        std::get<kEdgeRefPos>(*missing_in_edge) == std::get<kEdgeRefPos>(*missing_out_edge)) {
+      return std::make_tuple(std::get<kEdgeRefPos>(*missing_in_edge), std::get<kEdgeTypeIdPos>(*missing_in_edge),
+                             to_vertex, from_vertex);
+    }
+  }
+
+  const auto &from_edges = from_vertex->out_edges;
+  const auto &to_edges = to_vertex->in_edges;
+
+  auto it = std::find_if(from_edges.begin(), from_edges.end(), [&](const auto &from_entry) {
+    const auto &from_edge = std::get<kEdgeRefPos>(from_entry);
+    return std::any_of(to_edges.begin(), to_edges.end(), [&](const auto &to_entry) {
+      const auto &to_edge = std::get<kEdgeRefPos>(to_entry);
+      return index_iterator_->edge->gid == from_edge.ptr->gid && from_edge.ptr->gid == to_edge.ptr->gid;
+    });
+  });
```

Comment:
Why do we have here O(n^2) complexity?

Shouldn't we just iterate over all from_edges and over all to_edges and then just check they are same? This should be O(N) complexity in general

Reply:
I agree. Rewrote it